### PR TITLE
[fault tolerance] Add local checkpointing support

### PR DESCRIPTION
**Pull request by ananthsub**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?
- Add helpers for integrating NVRx local checkpointing with NeMo
- Bump NVRx version to 0.3 for reinstall

**Collection**: [Note which collection this PR will affect]

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

- You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: nemo/collections/llm/recipes/log/default.py [hunk 1]
```diff
@@ -67,8 +71,37 @@ def default_log(
...
 def default_resume(resume_if_exists=True, resume_ignore_no_checkpoint=True) -> Config[nl.AutoResume]:
+    """Factory function to configure AutoResume."""
     return Config(
         nl.AutoResume,
         resume_if_exists=resume_if_exists,
         resume_ignore_no_checkpoint=resume_ignore_no_checkpoint,
     )
+
+
+def get_global_step_from_global_checkpoint_path(path: Union[str, Path]) -> int:
```
Comment (hemildesai) [inline nemo/collections/llm/recipes/log/default.py:82]: Why is this in recipes? Looks like a util function.

File: nemo/lightning/pytorch/local_ckpt.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+import torch
+from lightning_fabric.plugins import CheckpointIO
+from lightning_fabric.utilities.types import _PATH
+from megatron.core.dist_checkpointing.tensor_aware_state_dict import MCoreTensorAwareStateDict
+from nvidia_resiliency_ext.checkpointing.local.base_state_dict import TensorAwareStateDict
+from nvidia_resiliency_ext.checkpointing.local.ckpt_managers.base_manager import BaseCheckpointManager
+from nvidia_resiliency_ext.checkpointing.local.ckpt_managers.local_manager import LocalCheckpointManager
+from nvidia_resiliency_ext.checkpointing.local.replication.strategies import LazyCliqueReplicationStrategy
+from nvidia_resiliency_ext.fault_tolerance.dict_utils import dict_list_map_inplace
+from nvidia_resiliency_ext.ptl_resiliency.local_checkpoint_callback import (
```
Comment (hemildesai) [inline nemo/lightning/pytorch/local_ckpt.py:29]: Does this need to be import guarded?

File: nemo/collections/llm/recipes/log/default.py (hunk 1 above)
Comment (ananthsub) [reply nemo/collections/llm/recipes/log/default.py:82]: this file defines the default filename convention in L59, so i've added this util function here as a convenience. 
since the filename is still ultimately configurable by the user, there's no generic solution we can provide unless we read the checkpoint data and get the step

File: nemo/lightning/pytorch/local_ckpt.py (hunk 2 above)
Comment (ananthsub) [reply nemo/lightning/pytorch/local_ckpt.py:29]: i'll add the import guard

File: nemo/collections/llm/recipes/log/default.py (hunk 1 above)
Comment (hemildesai) [reply nemo/collections/llm/recipes/log/default.py:82]: It doesn't seem to be used anywhere.

Comment (ananthsub) [reply nemo/collections/llm/recipes/log/default.py:82]: it's a utility function that users can pass to `update_trainer_local_checkpoint_io` if they use the filename format.

**Review by hemildesai — APPROVED** [review]
